Issue #2444: Load custom words spell check dictionary from Redmine wiki page

dmurphy1 opened the pull request:
This PR is for Redmine issue [#2444](https://pm.savaslabs.com/issues/2444) and improves the custom dictionary management for Sumac.

This PR does the following:

- Updates `config.example.yml` removing the `custom_words` key/value pair and replacing it with key/value pairs that point to a Redmine wiki page which will serve as the new custom words spellcheck dictionary.
- Makes updates to configure PSpell after the Redmine client has been initialized, so that `$this->redmineClient` is available
- Updates `configurePSpell()` to pull the list of custom words from the Redmine wiki.

To test locally:

- In your local instance of Redmine, add a new "Dictionary" wiki page to the Savaslabs project and enter custom words, one per line.
- Copy the new `spellcheck` key/value pairs from the `config.example.yml` file to your local `config.yml` file, and populate `wiki_page_name` with the name of the new wiki page you created
- Run a sync using `--slack-notify` - you should **not** be notified about typos for words in your local Redmine dictionary.

Deployment notes:

- When deploying, a dictionary wiki will need to be created in the production Redmine instance. The Sumac production `config.yml` will also need to be updated.

chrisarusso commented:
Excited for this one!

kostajh commented on src/Sumac/Console/Command/SyncCommand.php:
```diff
@@ -128,8 +128,19 @@ protected function configure()
      */
     private function configurePSpell()
     {
+        // Retrieve Redmine spelling dictionary wiki path.
```
We should check if these two fields are set first with `isset()`, and if not, return.

kostajh commented on src/Sumac/Console/Command/SyncCommand.php:
```diff
@@ -128,8 +128,19 @@ protected function configure()
+
+        // Load the wiki.
+        $wikiObject = new Redmine\Api\Wiki($this->redmineClient);
+        $wiki_page = $wikiObject->show($wiki_project_name, $wiki_page_name);
```
Could you please wrap this in a try/catch statement, log an error if the page doesn't exist, and abandon pspell configuration if so?

kostajh commented on src/Sumac/Console/Command/SyncCommand.php:
```diff
@@ -128,8 +128,19 @@ protected function configure()
+        $wiki_page = $wikiObject->show($wiki_project_name, $wiki_page_name);
+
+        // Populate words to ignore. The Redmine wiki uses "\r\n" for new lines.
+        $words_to_ignore = explode("\r\n", $wiki_page['wiki_page']['text']);
```
IMO it would be good to check if this is an array before proceeding.

dmurphy1 commented:
Hi @kostajh - please see my additions in 17eba1f. I updated the README, added checks for misconfigurations, and log warnings to the console in those cases. A couple specific notes:

- `$wiki_page = $wikiObject->show($wiki_project_name, $wiki_page_name);` returns `false` if no wiki is found, it does not throw an error - so I didn't use try/catch
- Since `explode` expects a string, I went with `if (!is_string($wiki_page['wiki_page']['text']))`
- As a result, `words_to_ignore` should always be an array as a result of `explode` always being run on a string. But just in case I added the `if (!is_array($words_to_ignore))` check, but I didn't log an error in that case.

kostajh commented:
@dmurphy1 I added a few more lines to sort the dictionary for us on each run. This should make it more maintainable. If it looks good, I can merge and deploy this.

dmurphy1 commented:
Hi @kostajh - nice additional, I think sorting the dictionary will be really helpful. When testing locally, I found that it sorted so all capitalized came before all lowercase words - which I think is fine but I just wanted to alert you. Here's what I got locally after sorting:

```
HEADER!
1
3
Anne
Dan
Kosta
alphabet
rad
test
zoo
```

I think this is good to merge and deploy! Nice work 👍 !

kostajh commented:
@dmurphy1 good catch! I switched to `natcasesort()` so that e.g. "Zoo" and "zoo" appear one line after another.
